src/partition/repart.c

```diff
@@ -3084,79 +3266,274 @@ static int partition_encrypt(
                 r = cryptsetup_add_token_json(cd, v);
                 if (r < 0)
                         return log_error_errno(r, "Failed to add TPM2 JSON token to LUKS2 header: %m");
+
+                passphrase = base64_encoded;
+                passphrase_size = strlen(base64_encoded);
 #else
                 return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
                                        "Support for TPM2 enrollment not enabled.");
 #endif
         }
 
-        r = sym_crypt_activate_by_volume_key(
+        r = sym_crypt_reencrypt_init_by_passphrase(
                         cd,
-                        dm_name,
-                        volume_key,
-                        volume_key_size,
-                        arg_discard ? CRYPT_ACTIVATE_ALLOW_DISCARDS : 0);
+                        NULL,
+                        passphrase,
+                        passphrase_size,
+                        CRYPT_ANY_SLOT,
+                        0,
+                        sym_crypt_get_cipher(cd),
+                        sym_crypt_get_cipher_mode(cd),
+                        &(struct crypt_params_reencrypt) {
+                                .mode = CRYPT_REENCRYPT_ENCRYPT,
+                                .direction = CRYPT_REENCRYPT_BACKWARD,
+                                .resilience = "datashift",
+                                .data_shift = LUKS2_METADATA_SIZE / (2 * context->sector_size),
+                                .luks2 = &luks_params,
+                                .flags = CRYPT_REENCRYPT_MOVE_FIRST_SEGMENT,
+                        });
```

Comment:
Ok, this goes to me and I need to improve API documentation for LUKS2 reencryption (this PR seems to be first user outside cryptsetup project).

Anyway, what crypt_reencrypt_init actually does is LUKS2 metadata initialization for later encryption process (and it also moves initial data segment backwards to spare area at the end of data device in case of CRYPT_REENCRYPT_MOVE_FIRST_SEGMENT flag. So mind that there has to be at least 2*data_shift spare space at the end of the image).

The real data encryption process is executed via crypt_reencrypt_run() API and I don't see the symbol anywhere in this PR :)

So if you aim to really encrypt data in image file, there's simplified steps:

1) ``crypt_format()`` placeholder header, with data device set to image file
2) Add a keyslot
3) ``crypt_reencrypt_init...(with flags containing CRYPT_REENCRYPT_INITIALIZE_ONLY and CRYPT_REENCRYPT_MOVE_FIRST_SEGMENT)`` this moves the initial data segment and prepares LUKS2 reencrypt metadata for later use
4) ``crypt_free(cd)``
5) ``crypt_init(image file)``
6) ``crypt_header_restore()`` to put LUKS2 header created in step 1 through 3 in the head of image file (inital data segment was moved in step 3))
7) ``crypt_reencrypt_init...(w/o CRYPT_REENCRYPT_INITIALIZE_ONLY flag)`` to load reencryption context for later use.
8) ``crypt_reencrypt_run()`` this actually encrypts data in image file and result is ordinary LUKS2 device (with no reencryption metadata)